network/common/src/main/java/org/apache/spark/network/buffer/WrappedLargeByteBuffer.java

```diff
@@ -0,0 +1,252 @@
+import java.nio.channels.WritableByteChannel;
+import java.util.Arrays;
+import java.util.List;
+
+public class WrappedLargeByteBuffer implements LargeByteBuffer {
+
+  //only public for tests
+  public final ByteBuffer[] underlying;
+
+  private final long size;
+  private long _pos;
+  private int currentBufferIdx;
+  private ByteBuffer currentBuffer;
+
+
+  public WrappedLargeByteBuffer(ByteBuffer[] underlying) {
+    this(underlying, findExpectedInitialPosition(underlying));
+  }
+
+  private static long findExpectedInitialPosition(ByteBuffer[] bufs) {
+    long sum = 0L;
+    for (ByteBuffer b: bufs) {
+      if (b.position() > 0) {
+        // this could still lead to a mix of positions half-way through buffers that
+        // would be inconsistent -- but we'll discover that in the constructor checks
+        sum += b.position();
+      } else {
+        break;
+      }
+    }
+    return sum;
+  }
+
+  private WrappedLargeByteBuffer(ByteBuffer[] underlying, long initialPosition) {
+    this.underlying = underlying;
+    long sum = 0L;
+    for (int i = 0; i < underlying.length; i++) {
+      ByteBuffer b = underlying[i];
+      long nextSum = sum + b.capacity();
+      int expectedPosition;
+      if (nextSum < initialPosition) {
+        expectedPosition = b.capacity();
+      } else if (sum > initialPosition) {
+        expectedPosition = 0;
+      } else {
+        expectedPosition = (int) (initialPosition - sum);
+      }
+      if (b.position() != expectedPosition) {
+        throw new IllegalArgumentException("ByteBuffer[" + i + "]:" + b + " was expected to have" +
+          " position = " + expectedPosition + " to be consistent with the overall " +
+          "initialPosition = " + initialPosition);
+      }
+      sum = nextSum;
+    }
+    _pos = initialPosition;
+    currentBufferIdx = 0;
+    currentBuffer = underlying[0];
+    size = sum;
+  }
+
+  @Override
+  public void get(byte[] dest, int offset, int length) {
+    if (length > remaining())
+      throw new BufferUnderflowException();
+    int moved = 0;
+    while (moved < length) {
+      int toRead = Math.min(length - moved, currentBuffer.remaining());
+      currentBuffer.get(dest, offset + moved, toRead);
+      moved += toRead;
+      updateCurrentBuffer();
+    }
+    _pos += moved;
+  }
+
+  @Override
+  public LargeByteBuffer rewind() {
+    while (currentBufferIdx > 0) {
+      if (currentBuffer != null) {
+        currentBuffer.rewind();
+      }
+      currentBufferIdx -= 1;
+      currentBuffer = underlying[currentBufferIdx];
+    }
+    currentBuffer.rewind();
+    _pos = 0;
+    return this;
+  }
+
+  @Override
+  public WrappedLargeByteBuffer deepCopy() {
+    ByteBuffer[] dataCopy = new ByteBuffer[underlying.length];
+    for (int i = 0; i < underlying.length; i++) {
+      ByteBuffer b = underlying[i];
+      dataCopy[i] = ByteBuffer.allocate(b.capacity());
+      int originalPosition = b.position();
+      b.position(0);
+      dataCopy[i].put(b);
```

Comment:
According to the `ByteBuffer` javadoc, this will only work if the current buffer is in its original state (or after a `rewind`). The interface documentation seems to imply that restriction does not exist.